Document Kotlin DSL (#3210)
spring-projects/spring-integration: src/reference/asciidoc/kotlin-dsl.adoc

        wireTap {
                    handle { println(it.payload) }
                }
        transform<String, String> { it.toUpperCase() }
    }
----
====

Such a global `integrationFlow()` function expects a lambda in builder style for a `KotlinIntegrationFlowDefinition` (a Kotlin wrapper for the `IntegrationFlowDefinition`) and produces a regular `IntegrationFlow` lambda implementation.
See more overloaded `integrationFlow()` variants below.

Many other scenarios require an `IntegrationFlow` to be started from source of data (e.g. `JdbcPollingChannelAdapter`, `JmsInboundGateway` or just an existing `MessageChannel`).
For this purpose Spring Integration Java DSL provides an `IntegrationFlows` factory with its bunch of overloaded `from()` methods.
This factory can be used in Kotlin as well:

====
[source, kotlin]
----
@Bean
fun flowFromSupplier() =
         IntegrationFlows.from<String>({ "bar" }) { e -> e.poller { p -> p.fixedDelay(10).maxMessagesPerPoll(1) } }
                 .channel { c -> c.queue("fromSupplierQueue") }
                 .get()
----
====

But unfortunately not all `from()` methods are compatible with Kotlin structures.

> To fix a gap, this project provides a Kotlin DSL around an `IntegrationFlows` factory.

Reviewer: ```suggestion
To fix the gap, this project provides a Kotlin DSL around an `IntegrationFlows` factory.
```

> It is done as a set of overloaded `integrationFlow()` functions.

Reviewer: ```suggestion
It is implemented as a set of overloaded `integrationFlow()` functions.
```

> With a consumer for a `KotlinIntegrationFlowDefinition` to declare the rest of the flow as an `IntegrationFlow` lambda to reuse the mentioned above experience and also avoid `get()` call in the end.

Reviewer: ```suggestion
With a consumer for a `KotlinIntegrationFlowDefinition`, to declare the rest of the flow as an `IntegrationFlow` lambda to reuse the mentioned above experience; unlike the Java DSL, it also avoids the need for a `get()` call at the end of the flow.
```

For example:

====
[source, kotlin]
----
@Bean
fun functionFlow() =
        integrationFlow<Function<String, String>>({ beanName("functionGateway") }) {
            transform<String, String> { it.toUpperCase() }
        }

@Bean
fun messageSourceFlow() =
        integrationFlow(MessageProcessorMessageSource { "testSource" },
                { poller { it.fixedDelay(10).maxMessagesPerPoll(1) } }) {
            channel { queue("fromSupplierQueue") }
        }
----
====

> In addition Kotlin extensions are provided for Java DSL API which needs some refinement for Kotlin structures.

Reviewer: ```suggestion
In addition, Kotlin extensions are provided for the Java DSL API which needs some refinement for Kotlin structures.
```
